src/goto-instrument/contracts/doc/user/contracts-loops.md

```diff
@@ -168,6 +168,40 @@ void array_wipe(__CPROVER_size_t len, char * array)
 }
 ```
 
+### Caution With Nested Loop
+Due to the nature of @ref contracts-assigns, we need to be aware of the non-deterministic value of the assigned target.
+In the example below, 
+```c
+const unsigned table[256] = {1, 2, 3, ..., 256};
+
+void nested_loop_example() {
+        unsigned t = 0;
+        for( j=0; j<18; j++ )
+        __CPROVER_assigns(t, j, k) // t and k need to be included in the outer loop
+        __CPROVER_loop_invariant(0 <= j && j <= 18 && t < 256) // t < 256 must be included in the outer loop invariant
+        __CPROVER_decreases(18 - j)
+        {
+            /*
+              during instrumentation, the assigns clause is equivalent to:
+              __CPROVER_havoc(t);
+              __CPROVER_havoc(j);
+            */
+            for( k=0; k<48; k++ )
+            __CPROVER_assigns(t, k)
+            __CPROVER_loop_invariant(0 <= k && k <= 48 && t < 256)
+            __CPROVER_decreases(48 - k)
+            {
+              /*
+                during loop abstraction, the assigns clause is equivalent to:
+                __CPROVER_havoc(t);
+                __CPROVER_havoc(k);
+              */
+                t = table[t];
+            }
+        }
+}
+```
+If `t < 256` is not included in the outer loop invariant, the inner loop invariant `t < 256` will immediately fail at loop entry because `__CPROVER_havoc(t)` at the outer loop will update `t` with a non-deterministic value which may be greater than 256. 
```

Comment:
```suggestion
If `t < 256` is not included in the outer loop invariant, the inner loop invariant `t < 256` will immediately fail at loop entry because, in the inductive step of the outer loop, the assigns target `t` of the outer loop will be a non-deterministic value which can be greater than 256. With the predicate `t<256` in the outer loop's invariants will restrict `t` to be less than `256` in the proof of the inductive step of the outer loop.
```